### PR TITLE
Oracle2bq notebook

### PR DESCRIPTION
**Pull request by somanishivam**
#223

### CODE REVIEW

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+   "metadata": {},
+   "source": [
+    "### Step 3.1 Common Parameters\n",
+    " \n",
+    "- PROJECT : GCP project-id\n",
+    "- REGION : GCP region\n",
+    "- GCS_STAGING_LOCATION : GCS staging location to be used for this notebook to store artifacts\n",
+    "- SUBNET : VPC subnet\n",
+    "- JARS : list of jars. For this notebook mysql connector and avro jar is required in addition with the dataproc template jars\n",
+    "- MAX_PARALLELISM : Parameter for number of jobs to run in parallel default value is 2"
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:196]: we should set MAX_PARALLELISM default value to `5`

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f8ac9a9f-1332-4c75-b6ca-bfb3b776671f",
+   "metadata": {},
+   "source": [
+    "# Step 6: Get Row Count of Tables and identify Partition Columns\n",
+    "This step uses PARTITION_THRESHOLD(default value is 1 million) parameter and any table having rows greater than PARTITION_THRESHOLD will be used for partitioned read based on Primary Keys\n",
+    " - NUMBER_OF_PARTITIONS : Number of partitioned reads default 10\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:394]: NUMBER_OF_PARTITIONS should not be hardcoded.
Refer to MySQLToSpanner notebook for latest code.

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 3]
```diff
@@ -0,0 +1,888 @@
...
+   "id": "b692a051-7555-42fc-af27-b05e461576f4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "with DB.connect() as conn:\n",
+    "    for table in ORACLETABLE_LIST:\n",
+    "        results = DB.execute(\"SELECT count(1) FROM {}\".format(table)).fetchall()\n",
+    "        if results[0][0]>int(PARTITION_THRESHOLD) and len(SQL_TABLE_PRIMARY_KEYS.get(table).split(\",\")[0])>0:\n",
+    "            column_list=SQL_TABLE_PRIMARY_KEYS.get(table).split(\",\")\n",
+    "            for column in column_list:\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:422]: Why for loop?
refer to MySQLtoSpanner Notebook.

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+    "                print(column)\n",
+    "                results_datatype = DB.execute(\"select DATA_TYPE from sys.all_tab_columns where TABLE_NAME = '{0}' and COLUMN_NAME = '{1}'\".format(table,column)).fetchall()      \n",
+    "                print(results_datatype[0][0])\n",
+    "                if results_datatype[0][0]==\"NUMBER\":\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:426]: Why only NUMBER ? What about int, bigint and other similar data types ?

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+   "execution_count": null,
+   "id": "4b40f634-1983-4267-a4c1-b072bf6d81ae",
+   "metadata": {
+    "scrolled": true,
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "wget https://repo1.maven.org/maven2/com/oracle/database/jdbc/ojdbc8/21.7.0.0/ojdbc8-21.7.0.0.jar\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:473]: Download only when jar is missing.

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "06735281-1c22-48d3-9b71-20f9ca99d04a",
+   "metadata": {
+    "scrolled": true,
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "! python ./setup.py bdist_egg --output=$PACKAGE_EGG_FILE"
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:487]: Should `python3` be used ?

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 7]
```diff
@@ -0,0 +1,888 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1f98914b-bd74-4d0e-9562-7019d504a25e",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "oracle_to_gcs_jobs = []\n",
+    "BIGQUERY_DATASET=\"oracle_dump\"\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:571]: hardcoding

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 8]
```diff
@@ -0,0 +1,888 @@
...
+    "BIGQUERY_MODE = \"append\"\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:572]: default mode should be overwrite.
Also specify other write modes commented, next to it

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 9]
```diff
@@ -0,0 +1,888 @@
...
+    "- [Build a Vertex AI Pipeline](https://cloud.google.com/vertex-ai/docs/pipelines/build-pipeline)\n",
+    "\n",
+    "This notebook is built to run a Vertex AI User-Managed Notebook using the default Compute Engine Service Account.  \n",
+    "Check the Dataproc Serverless in Vertex AI Pipelines tutorial linked above to learn how to setup a different Service Account.  \n",
+    "\n",
+    "#### Permissions\n",
+    "\n",
+    "Make sure that the service account used to run the notebook has the following roles:\n",
+    "\n",
+    "- roles/aiplatform.serviceAgent\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:52]: missing BQ permissions in the list.

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 10]
```diff
@@ -0,0 +1,888 @@
...
+    "\n",
+    "- [DataprocPySparkBatchOp reference](https://google-cloud-pipeline-components.readthedocs.io/en/google-cloud-pipeline-components-1.0.0/google_cloud_pipeline_components.experimental.dataproc.html)\n",
+    "- [Kubeflow SDK Overview](https://www.kubeflow.org/docs/components/pipelines/sdk/sdk-overview/)\n",
+    "- [Dataproc Serverless in Vertex AI Pipelines tutorial](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/ml_ops/stage3/get_started_with_dataproc_serverless_pipeline_components.ipynb)\n",
+    "- [Build a Vertex AI Pipeline](https://cloud.google.com/vertex-ai/docs/pipelines/build-pipeline)\n",
+    "\n",
+    "This notebook is built to run a Vertex AI User-Managed Notebook using the default Compute Engine Service Account.  \n",
+    "Check the Dataproc Serverless in Vertex AI Pipelines tutorial linked above to learn how to setup a different Service Account.  \n",
+    "\n",
+    "#### Permissions\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:84]: permissions are duplicated

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 11]
```diff
@@ -0,0 +1,888 @@
...
+    "- roles/aiplatform.serviceAgent\n",
+    "- roles/aiplatform.customCodeServiceAgent\n",
+    "- roles/storage.objectCreator\n",
+    "- roles/storage.objectViewer\n",
+    "- roles/dataproc.editor\n",
+    "- roles/dataproc.worker"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:97]: Add a section about feedback form and feedback email id before Step 1

**Review by shashank-google — CHANGES_REQUESTED** [review]

File: notebooks/oracle2bq/README.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+### Overview
+
+[OracleToBigQuery_vertex_pipeline_pyspark.ipynb](https://github.com/GoogleCloudPlatform/dataproc-templates/blob/oracle2bq-notebook/notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb) notebook solution is built on top of [Vertex AI Jupyter Notebook](https://cloud.google.com/vertex-ai/docs/tutorials/jupyter-notebooks) and [Google Cloud's Dataproc](https://cloud.google.com/dataproc/) tech stack provided by GCP.
+
+### Key Benifits
+
+* Automatically Generate list of tables from metadata. Alternatively, user should be able to supply list of tables.
+* Identify current primary key column name, and partitioned read properties.
+* Use partition reads if exceeds threshold.
```
Comment (shashank-google) [inline notebooks/oracle2bq/README.md:16]: `Automatically uses partitioned reads`

File: notebooks/oracle2bq/README.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+* Divides migration into batches and parallely migrates multiple table.
```
Comment (shashank-google) [inline notebooks/oracle2bq/README.md:17]: `multiple tables`

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 14]
```diff
@@ -0,0 +1,888 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "from google.cloud import bigquery\n",
+    "\n",
+    "# Construct a BigQuery client object.\n",
+    "client = bigquery.Client()\n",
+    "\n",
+    "for table in ORACLETABLE_LIST:\n",
+    "    results = client.query(\"select count(*) from {}.{}\".format(BIGQUERY_DATASET,table))\n",
```
Comment (shashank-google) [inline notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:828]: `count(*)` in BQ is expensive operation.
Can we get the row counts from information schema instead ?

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb (hunk 9 above)
Comment (shashank-google) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:52]: can you please test and ensure that this list of roles is correct, by defining a service account which has only these roles.

**Review by shashank-google — CHANGES_REQUESTED** [review]

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb (hunk 5 above)
Comment (somanishivam) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:482]: was getting error without them, only required one downloaded

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb (hunk 11 above)
Comment (somanishivam) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:76]: added

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 15]
```diff
@@ -0,0 +1,888 @@
...
+   "metadata": {
+    "scrolled": true,
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%bash\n",
+    "sudo mkdir -p /opt/oracle\n",
+    "cd /opt/oracle\n",
+    "sudo wget https://download.oracle.com/otn_software/linux/instantclient/instantclient-basic-linuxx64.zip\n",
```
[the remark this answers is not in the document]
Comment (somanishivam) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb:110]: noted, oracle installation guide - https://docs.oracle.com/en/database/oracle/oracle-database/21/lacli/install-instant-client-using-zip.html#GUID-D3DCB4FB-D3CA-4C25-BE48-3A1FB5A22E84

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 16]
```diff
@@ -0,0 +1,888 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fef65ec2-ad6b-407f-a993-7cdf871bba11",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%bash\n",
```
[the remark this answers is not in the document]
Comment (somanishivam) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb]: added

File: notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb [hunk 17]
```diff
@@ -0,0 +1,888 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "19ef1307-902e-4713-8948-b86084e19312",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# waititng some time for kernel to restart\n",
```
[the remark this answers is not in the document]
Comment (somanishivam) [reply notebooks/oracle2bq/OracleToBigQuery_notebook.ipynb]: commented

**Review by shashank-google — APPROVED** [review]
left one comment.
once it is tested feel free to merge

**Review by hhasija — APPROVED** [review]
LGTM

**Pull request merged**